File: scripts/translate-quiz-i18n.py
# -*- coding: utf-8 -*-
"""RU → EN/UZ quiz translation. One JSON file per question (no races), then merge."""
from __future__ import annotations

import json
import time
from concurrent.futures import ThreadPoolExecutor, as_completed
from pathlib import Path

from deep_translator import GoogleTranslator

ROOT = Path(__file__).resolve().parents[1]
SECTION_SRC = ROOT / "scripts/_generated/quiz-i18n-source.json"
LOGICAL_SRC = ROOT / "scripts/_generated/logical-quiz-i18n-source.json"
SECTION_PARTS = ROOT / "scripts/_generated/quiz-i18n-parts"
LOGICAL_PARTS = ROOT / "scripts/_generated/logical-i18n-parts"
SECTION_OUT = ROOT / "src/data/sectionQuizI18n.json"
LOGICAL_OUT = ROOT / "src/data/g7LogicalQuizI18n.json"

SEP = "\n|||\n"
WORKERS = 4


def load(path: Path, fallback):
    if not path.exists():
        return fallback
    return json.loads(path.read_text(encoding="utf-8"))


def save(path: Path, data) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(json.dumps(data, ensure_ascii=False, indent=2) + "\n", encoding="utf-8")


def is_complete(entry) -> bool:
    return bool(
        entry
        and isinstance(entry.get("questionEn"), str)
        and entry["questionEn"].strip()
        and isinstance(entry.get("questionUz"), str)
        and entry["questionUz"].strip()
        and isinstance(entry.get("choicesEn"), list)
        and len(entry["choicesEn"]) == 4
        and isinstance(entry.get("choicesUz"), list)
        and len(entry["choicesUz"]) == 4
    )


def fields(q: dict) -> list[str]:
    ch = list(q.get("choices") or [])
    while len(ch) < 4:
        ch.append("")
    return [q.get("question") or "", ch[0], ch[1], ch[2], ch[3], q.get("explanation") or ""]


def split_parts(text: str) -> list[str]:
    parts = [p.strip() for p in (text or "").split("|||")]
    while len(parts) < 6:
        parts.append("")
    return parts[:6]


def translate_joined(parts: list[str], target: str, retries: int = 5) -> list[str]:
    joined = SEP.join(parts)
    if not joined.replace("|", "").strip():
        return ["", "", "", "", "", ""]
    translator = GoogleTranslator(source="ru", target=target)
    last_err = None
    for attempt in range(retries):
        try:
            raw = translator.translate(joined[:4500]) or ""
            return split_parts(raw)
        except Exception as e:  # noqa: BLE001
            last_err = e
            time.sleep(1.1 * (attempt + 1))
    raise RuntimeError(str(last_err))


def translate_item(q: dict) -> dict:
    parts = fields(q)
    en = translate_joined(parts, "en")
    uz = translate_joined(parts, "uz")
    return {
        "questionEn": en[0],
        "questionUz": uz[0],
        "choicesEn": en[1:5],
        "choicesUz": uz[1:5],
        "explanationEn": en[5],
        "explanationUz": uz[5],
    }


def part_path(parts_dir: Path, key: str) -> Path:
    safe = key.replace("/", "_").replace("\\", "_")
    return parts_dir / f"{safe}.json"


def merge_parts(parts_dir: Path, out_path: Path) -> int:
    merged = load(out_path, {})
    if not parts_dir.exists():
        save(out_path, merged)
        return len(merged)
    for f in parts_dir.glob("*.json"):
        try:
            entry = json.loads(f.read_text(encoding="utf-8"))
        except Exception:  # noqa: BLE001
            continue
        if is_complete(entry):
            merged[f.stem] = entry
    # also keep already-good keys from out
    save(out_path, {k: v for k, v in merged.items() if is_complete(v)})
    return len(load(out_path, {}))


def process(src: Path, parts_dir: Path, out_path: Path, key_field: str) -> None:
    parts_dir.mkdir(parents=True, exist_ok=True)
    source = load(src, [])
    # Seed parts from existing out file
    existing = load(out_path, {})
    for key, entry in existing.items():
        if is_complete(entry):
            save(part_path(parts_dir, key), entry)

    pending = []
    for q in source:
        key = q.get(key_field) or q["id"]
        p = part_path(parts_dir, key)
        if p.exists():
            try:
                if is_complete(json.loads(p.read_text(encoding="utf-8"))):
                    continue
            except Exception:  # noqa: BLE001
                pass
        pending.append(q)

    print(f"{out_path.name}: total={len(source)} pending={len(pending)} workers={WORKERS}", flush=True)
    done = 0

    def one(q: dict):
        key = q.get(key_field) or q["id"]
        entry = translate_item(q)
        save(part_path(parts_dir, key), entry)
        return key

    if pending:
        with ThreadPoolExecutor(max_workers=WORKERS) as ex:
            futures = [ex.submit(one, q) for q in pending]
            for fut in as_completed(futures):
                try:
                    fut.result()
                    done += 1
                    if done % 10 == 0 or done == len(pending):
                        n = merge_parts(parts_dir, out_path)
                        print(f"  parts {done}/{len(pending)} merged={n}", flush=True)
                except Exception as e:  # noqa: BLE001
                    print(f"  FAIL: {e}", flush=True)
                    time.sleep(0.5)

    n = merge_parts(parts_dir, out_path)
    print(f"  final merged={n}", flush=True)


def main() -> None:
    process(SECTION_SRC, SECTION_PARTS, SECTION_OUT, "id")
    if LOGICAL_SRC.exists():
        process(LOGICAL_SRC, LOGICAL_PARTS, LOGICAL_OUT, "templateKey")
    print("done", flush=True)


if __name__ == "__main__":
    main()

File: scripts/translate-quiz-i18n.mjs
/**
 * Массовый перевод MCQ RU → EN + UZ через OpenAI.
 * Usage: node --env-file=.env scripts/translate-quiz-i18n.mjs
 * Resume-safe: дописывает src/data/sectionQuizI18n.json и g7LogicalQuizI18n.json
 */
import fs from 'node:fs'
import path from 'node:path'

const ROOT = path.resolve(import.meta.dirname, '..')

function loadEnvFile() {
  const envPath = path.join(ROOT, '.env')
  if (!fs.existsSync(envPath)) return
  for (const line of fs.readFileSync(envPath, 'utf8').split(/\r?\n/)) {
    const m = line.match(/^\s*([A-Za-z_][A-Za-z0-9_]*)\s*=\s*(.*)$/)
    if (!m) continue
    let v = m[2].trim()
    if ((v.startsWith('"') && v.endsWith('"')) || (v.startsWith("'") && v.endsWith("'"))) {
      v = v.slice(1, -1)
    }
    if (!(m[1] in process.env)) process.env[m[1]] = v
  }
}

loadEnvFile()

const API_KEY = process.env.OPENAI_API_KEY
if (!API_KEY) {
  console.error('OPENAI_API_KEY missing')
  process.exit(1)
}

const MODEL = process.env.QUIZ_I18N_MODEL || 'gpt-4o-mini'
const BATCH = Number(process.env.QUIZ_I18N_BATCH || 8)
const CONCURRENCY = Number(process.env.QUIZ_I18N_CONCURRENCY || 3)

const SECTION_OUT = path.join(ROOT, 'src/data/sectionQuizI18n.json')
const LOGICAL_OUT = path.join(ROOT, 'src/data/g7LogicalQuizI18n.json')
const SECTION_SRC = path.join(ROOT, 'scripts/_generated/quiz-i18n-source.json')
const LOGICAL_SRC = path.join(ROOT, 'scripts/_generated/logical-quiz-i18n-source.json')

function loadJson(file, fallback) {
  if (!fs.existsSync(file)) return fallback
  return JSON.parse(fs.readFileSync(file, 'utf8'))
}

function saveJson(file, data) {
  fs.writeFileSync(file, `${JSON.stringify(data, null, 2)}\n`, 'utf8')
}

function isComplete(entry) {
  return (
    entry &&
    typeof entry.questionEn === 'string' &&
    entry.questionEn.trim() &&
    typeof entry.questionUz === 'string' &&
    entry.questionUz.trim() &&
    Array.isArray(entry.choicesEn) &&
    entry.choicesEn.length === 4 &&
    Array.isArray(entry.choicesUz) &&
    entry.choicesUz.length === 4
  )
}

async function translateBatch(items) {
  const payload = items.map((q) => ({
    id: q.id,
    question: q.question,
    choices: q.choices,
    explanation: q.explanation || '',
  }))

  const system = `You translate school chemistry multiple-choice questions from Russian to English and Uzbek (Latin script, Uzbekistan).
Return ONLY valid JSON object: { "items": [ ... ] }.
For each input item output exactly:
{
  "id": "<same id>",
  "questionEn": "...",
  "questionUz": "...",
  "choicesEn": ["a","b","c","d"],
  "choicesUz": ["a","b","c","d"],
  "explanationEn": "...",
  "explanationUz": "..."
}
Rules:
- Keep chemical formulas, symbols, numbers, subscripts, and proper names accurate (H2O, Al-Kindi/Al-Kindiy, etc.).
- Preserve choice ORDER exactly (correct answer stays at the same index).
- Uzbek: natural school language, Latin alphabet (o', g', sh, ch).
- explanation may be empty string if source explanation is empty.
- No markdown fences.`

  const res = await fetch('https://api.openai.com/v1/chat/completions', {
    method: 'POST',
    headers: {
      Authorization: `Bearer ${API_KEY}`,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      model: MODEL,
      temperature: 0.2,
      response_format: { type: 'json_object' },
      messages: [
        { role: 'system', content: system },
        { role: 'user', content: JSON.stringify({ items: payload }) },
      ],
    }),
  })

  if (!res.ok) {
    const text = await res.text()
    throw new Error(`OpenAI ${res.status}: ${text.slice(0, 400)}`)
  }
  const data = await res.json()
  const content = data.choices?.[0]?.message?.content
  if (!content) throw new Error('Empty completion')
  const parsed = JSON.parse(content)
  const list = Array.isArray(parsed.items) ? parsed.items : Array.isArray(parsed) ? parsed : null
  if (!list) throw new Error('Bad JSON shape')
  return list
}

function chunk(arr, size) {
  const out = []
  for (let i = 0; i < arr.length; i += size) out.push(arr.slice(i, i + size))
  return out
}

async function mapPool(items, limit, fn) {
  let i = 0
  const results = []
  const workers = Array.from({ length: limit }, async () => {
    while (i < items.length) {
      const idx = i++
      results[idx] = await fn(items[idx], idx)
    }
  })
  await Promise.all(workers)
  return results
}

async function processFile(srcPath, outPath, keyField) {
  const source = loadJson(srcPath, [])
  const out = loadJson(outPath, {})
  const pending = source.filter((q) => !isComplete(out[q[keyField] || q.id]))
  console.log(`${path.basename(outPath)}: total=${source.length} pending=${pending.length}`)
  if (pending.length === 0) return out

  const batches = chunk(pending, BATCH)
  let done = 0
  await mapPool(batches, CONCURRENCY, async (batch) => {
    let attempt = 0
    for (;;) {
      try {
        const translated = await translateBatch(batch)
        const byId = new Map(translated.map((t) => [t.id, t]))
        for (const q of batch) {
          const key = q[keyField] || q.id
          const t = byId.get(q.id)
          if (!t || !Array.isArray(t.choicesEn) || t.choicesEn.length !== 4) {
            throw new Error(`Missing/bad translation for ${q.id}`)
          }
          out[key] = {
            questionEn: String(t.questionEn || '').trim(),
            questionUz: String(t.questionUz || '').trim(),
            choicesEn: t.choicesEn.map(String),
            choicesUz: t.choicesUz.map(String),
            explanationEn: String(t.explanationEn || '').trim(),
            explanationUz: String(t.explanationUz || '').trim(),
          }
        }
        done += batch.length
        saveJson(outPath, out)
        console.log(`  saved ${done}/${pending.length}`)
        return
      } catch (err) {
        attempt++
        console.warn(`  batch fail (try ${attempt}):`, err.message || err)
        if (attempt >= 4) throw err
        await new Promise((r) => setTimeout(r, 1500 * attempt))
      }
    }
  })
  return out
}

const section = await processFile(SECTION_SRC, SECTION_OUT, 'id')
console.log('section complete', Object.keys(section).length)

if (fs.existsSync(LOGICAL_SRC)) {
  const logical = await processFile(LOGICAL_SRC, LOGICAL_OUT, 'templateKey')
  console.log('logical complete', Object.keys(logical).length)
}

console.log('done')
